Mathlib/Data/Bitvec/Core.lean

```diff
@@ -310,48 +311,84 @@ attribute [local simp] Nat.zero_add Nat.add_zero Nat.one_mul Nat.mul_one Nat.zer
 
 -- mul_left_comm
 theorem toNat_append {m : ℕ} (xs : Bitvec m) (b : Bool) :
-    Bitvec.toNat (xs++ₜb ::ᵥ nil) = Bitvec.toNat xs * 2 + Bitvec.toNat (b ::ᵥ nil) :=
-  by
+    Bitvec.toNat (xs++ₜb ::ᵥ nil) = Bitvec.toNat xs * 2 + Bitvec.toNat (b ::ᵥ nil) := by
   cases' xs with xs P
-  simp [bits_to_nat_to_list]; clear P
-  unfold bits_to_nat List.foldl
+  simp [bitsToNat_toList]; clear P
+  unfold bitsToNat List.foldl
   -- generalize the accumulator of foldl
   generalize h : 0 = x;
-  conv in add_lsb x b => rw [← h]; clear h
+  conv in addLsb x b =>
+    rw [← h];
   simp
   induction' xs with x xs generalizing x
   · simp
-    unfold List.foldl add_lsb
+    unfold addLsb
+    simp
     simp [Nat.mul_succ]
   · simp
-    apply xs_ih
+    rename_i b' _
+    rw [←h]
+    have s₁: addLsb (List.foldl addLsb (addLsb 0 b') xs) b =
+      List.foldl addLsb (addLsb 0 b') xs + List.foldl addLsb (addLsb 0 b') xs + cond b 1 0 := by
+        rfl
+    have _ :  addLsb 0 b = 0 + 0 + cond b 1 0  := by
+      rfl
+    have s₃ : 0 + 0 + cond b 1 0 = cond b 1 0 := by simp
+    have s₄ : addLsb 0 b = cond b 1 0 := by
+      apply s₃
+    have s₅ : ∀ n : Nat, n + n = n * 2 := by
+      intro n
+      conv =>
+        lhs
+        rw[←Nat.mul_one n,←Nat.mul_add]
+    have s₆: (List.foldl addLsb (addLsb 0 b') xs)
+      + (List.foldl addLsb (addLsb 0 b') xs) + cond b 1 0 =
+     (List.foldl addLsb (addLsb 0 b') xs) * 2 + cond b 1 0 := by
+      rw [s₅ _]
+    have s₇ : addLsb (List.foldl addLsb (addLsb 0 b') xs) b
+      = (List.foldl addLsb (addLsb 0 b') xs) * 2 + cond b 1 0 := by
+      rw[s₁, s₆]
+    have s₈: addLsb (List.foldl addLsb (addLsb 0 b') xs) b =
+     (List.foldl addLsb (addLsb 0 b') xs) * 2 + addLsb 0 b := by
+     rw[s₇,s₄]
+    rw [Nat.add_comm]
+    exact s₈
```

Comment:
What happened here? Why did the mathport output not work?

Reply:
the name `xs_ih` was one of the hidden names that the lean3 proof generated. The rest of it is because inductive hypothesis seemed to unfold differently and not `apply ih` would not work on it. The `unfold List.foldl` unfolded the `ih`, but the goal remained untouched.

Reply:
`simp only` or `simp` did not solve it either at the line `apply xs_ih` after fixing the above issues. I had to manually squash these equations together because lean4 `simp` was not identifying the patterns and applying the lemmas of `Nat.add_comm` and `Nat.mul_add` etc.  So I chose to manually apply them.

Reply:
> the name `xs_ih` was one of the hidden names that the lean3 proof generated.

Then you should give it a non-hidden name in the `induction` line, assuming it's being used.

Reply:
Yeah it is not being used in any case. the boolean b' is being used though. I introduced it. I had also introduced `tail_ih` to use the mathlib3port version, but it was pointless. So I removed it.

Reply:
Isn't the boolean `b` the `x` that is already introduced by the induction line?

Reply:
There is also an `x : Nat`. Hence to disambiguate, I use `b`

Reply:
The `x` that `induction` is supposed to be introducing is hidden as `x^{\dagger}`

Reply:
@eric-wieser I made a correction. When comparing with the mathlib3port proof, I noticed I had removed a `clear h`. I introduced it and fixed the proof again.

Reply:
Resolving, I've deleted this proof in favor of the mathlib3 one